[18기_이종미] Spring DB 모델링과 JPA 미션 제출합니다.

jongmee opened the pull request:
스프링 2주차 스터디 미션 제출합니다.

jongmee committed:
:tada: 프로젝트 생성하기

jongmee committed:
:sparkles: 의존성 추가하여 build.gradle 업데이트하기

jongmee committed:
:sparkles: 도메인 생성하기

jongmee committed:
:sparkles: 각 Entity 별 Jpa Repository 생성하기

jongmee committed:
:sparkles: 회원 도메인에 전화번호 속성 추가하기

jongmee committed:
:white_check_mark: PostRepository 테스트하기

jongmee committed:
:wrench: 불필요한 logging 설정 삭제하기

- 쿼리가 2번씩 출력되어 logging 설정을 삭제하였다

jongmee committed:
:memo: 2주차 리드미 작성하기

mirageoasis commented on src/main/java/com/daagn/clonestudy/BaseTimeEntity.java:
```diff
@@ -0,0 +1,20 @@
+  @CreatedDate
+  @Column(updatable = false, nullable = false)
+  private LocalDateTime createdDate;
+
```
마지막 수정 시간도 추가해주면 좋아요~

mirageoasis commented on src/main/java/com/daagn/clonestudy/Chatting.java:
```diff
@@ -0,0 +1,36 @@
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  private Long id;
+
+  @JoinColumn(nullable = false)
```
개인적으로 나중에 디비를 볼 때를 위해서 JoinColumn에서 name을 post_id이런식으로 해보는 것도 좋다고 생각해요

mirageoasis commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,14 @@
+spring:
```
마찬가지로 좋지 않아요~ yml 파일~
[링크](https://velog.io/@kshired/Spring-yml-%EC%84%A4%EC%A0%95-%ED%8C%8C%EC%9D%BC-%EA%B4%80%EB%A6%AC%ED%95%98%EA%B8%B0)
한번 이 블로그 글을 보는 것도 나쁘지 않아요

mirageoasis commented on src/test/java/com/daagn/clonestudy/PostRepositoryTest.java:
```diff
@@ -0,0 +1,45 @@
+    Post 게시글3 = Post.builder().title("테스트 제목3").price(15000).address("테스트 거래 주소3").isAuction(false).writer(테스트유저2).build();
+
+    // when
+    memberRepository.save(테스트유저1);
```
saveAll이라는 메서드도 있으니 알아보세유~

jongmee replied:
오 보기 깔끔하게 리팩토링할 수 있겠네요! 감사합니당 🙇🏻‍♀️

chanhyeokseo commented on README.md:
```diff
@@ -1,2 +1,142 @@
+  3. <b>Post</b><br>
+  가격 제안을 받는 여부, 구매 상태(판매 중, 예약 중, 거래 완료), 거래 희망 장소, 작성자의 pk 등 당근마켓의 게시물 요소를 반영하였다.
+  4. <b>PostImage</b><br>
+  한 게시물 당 8개까지 이미지를 올릴 수 있다. 게시물과 이미지가 Many-to-one 관계이므로 게시물 이미지 테이블을 생성하였다.
```
대표 이미지 (또는 이미지 순서)는 어떻게 구현할 수 있을까요?

jongmee replied:
PostImage에 1) 대표 이미지를 나타내는 flag를 추가하거나 2) 가장 첫 번째 이미지를 대표 이미지로 생각해서 반환하면 될 것 같아요! 당근 마켓 UI도 첫 번째 이미지를 '대표 이미지' 표시를 해주는데, 저는 두 번째로 구현할 것 같습니다 😁

chanhyeokseo commented on README.md:
```diff
@@ -1,2 +1,142 @@
+  1. <b>Member</b><br>
+  매너온도, 재거래희망률, 응답률을 가진다.
+  2. <b>Review</b><br>
+  리뷰 쓴 유저와 리뷰 받은 유저를 Foreign key로 저장한다. is_buyer라는 속성으로 리뷰 남긴 이가 구매자인지 판매자인지 구별한다.
```
어떤 거래 후기가 몇 점을 주었는지 확인할 수 없는 것 같아요.
의도적으로 그렇게 설계하신건가요?

jongmee replied:
매너 온도 값을 업데이트 하려면 긍정 매너 평가 혹은 부정 매너 평가로 판단해야 하는데, 그 부분을 고려하지 못했네요 🥲 DB 수정에 반영하겠습니다! 좋은 리뷰 감사합니다 🙇🏻‍♀️

chanhyeokseo reviewed (commented):
종미님 미션 수고하셨습니다!! 남은 연휴 잘 보내시고 스터디때 보아요~

jongmee committed:
:sparkles: 2주차 추가 과제 1, 2번 문제 해결하기

jongmee committed:
:memo: README.md 수정

jongmee committed:
:recycle: 테스트 코드 리팩토링하기

jongmee committed:
:sparkles: Review Entity를 MannerEvaluation Entity로 변경하고 속성 추가하기

- 코드 리뷰를 통해 리뷰 엔티티에 평가점수에 대응되는 속성이 없다는 것을 알게 되었다. 매너 온도에 거래 후기가 영향을 주지만, 리뷰 내용만 보고 긍정적 리뷰인지 부정적 리뷰인지 판단하는 알고리즘을 도입하기 어려울 것 같아서 리뷰 대신 매너 평가를 도입하였다. (실제 당근 마켓에 매너 온도에 영향을 주는 한 기능이다. 또한 당근 마켓의 거래 후기에는 별점이 없다.)

itsme-shawn commented on README.md:
```diff
@@ -1,2 +1,270 @@
+    List<Post> 저장된_게시글들 = postRepository.findAll();
+
+    // then
+    assertThat(저장된_게시글들).hasSize(3)
+        .extracting(Post::getTitle)
+        .contains("테스트 제목1", "테스트 제목2", "테스트 제목3");
```
extracting 이라는 함수 처음봤는데 배워갑니다

itsme-shawn commented on README.md:
````diff
@@ -1,2 +1,270 @@
+     ```
+     - MyRepository에 findAllByName과 같이 메서드를 만들면 동적으로 Proxy 인스턴스에 생성된다.
+2. Data jpa를 찾다보면 SimpleJpaRepository에서  <b>entity manager</b>를 생성자 주입을 통해서 주입 받는다. 근데 <b>싱글톤</b> 객체는 한번만 할당을  받는데, 한번 연결 때 마다 생성이 되는 entity manager를 생성자 주입을 통해서 받는 것은 수상하지 않는가? 어떻게 되는 것일까?<br>
+➡️ 주입 받는 Entity Manamger 역시 프록시 객체이다. Entity Manager의 메서드를 호출하면 SharedEntityManagerInvocationHandler를 호출하여 메서드를 실행한다. 현재 데이터베이스 트랜잭션과 관련된 실제 EntityManager의 메서드를 호출할 수 있다. 싱글톤은 유지하면서 동시성 이슈를 해결할 수 있다.
````
이것도 몰랐는데 알아갑니다 👍

itsme-shawn commented on README.md:
```diff
@@ -1,2 +1,270 @@
+  한 게시물 당 8개까지 이미지를 올릴 수 있다. 게시물과 이미지가 Many-to-one 관계이므로 게시물 이미지 테이블을 생성하였다.
+  5. <b>Purchase</b><br>
+  구매자와 게시글을 Foreign key로 저장하고 구매 확정 여부를 저장한다.
+  6. <b>Chatting</b><br>
+  한 게시글 보고 여러 유저가 게시자에게 채팅할 수 있다. 따라서 채팅방 테이블에는 게시글, 채팅 한 유저, 채팅 받은 유저의 Foreign key를 저장한다.
```
Chatting 방에는 Member FK 를 2개 두는 방식을 선택하셨군요
저는 Member 와 Chatting 은 ManyToMany 이므로 저는 중간에 연결테이블을 둬서 OneToMany 로 풀어내는 방식을 택했습니다.

jongmee replied:
그 방식이면 단체 채팅도 할 수 있겠네요 ! 좋은 방법인 것 같습니다

itsme-shawn reviewed (commented):
안녕하세요
이번 주 코드리뷰를 맡은 최현수입니다

README 작성이 너무 깔끔하시네요

그리고 추가과제 부분도 정리 해주셔서 몰랐던 부분 채워갑니다

한 주간 고생하셨습니다

yoonsseo commented on README.md:
```diff
@@ -1,2 +1,270 @@
+  가격 제안을 받는 여부, 구매 상태(판매 중, 예약 중, 거래 완료), 거래 희망 장소, 작성자의 pk 등 당근마켓의 게시물 요소를 반영하였다.
+  4. <b>PostImage</b><br>
+  한 게시물 당 8개까지 이미지를 올릴 수 있다. 게시물과 이미지가 Many-to-one 관계이므로 게시물 이미지 테이블을 생성하였다.
+  5. <b>Purchase</b><br>
+  구매자와 게시글을 Foreign key로 저장하고 구매 확정 여부를 저장한다.
```
구매 확정을 어떻게 처리할 지 고민했었는데 구매 확정용 테이블을 따로 두는 방법도 있겠네용!!👍👍

yoonsseo commented on README.md:
```diff
@@ -1,2 +1,270 @@
+
+    // then
+    assertThat(저장된_게시글들).hasSize(3)
+        .extracting(Post::getTitle)
```
저도 배워갑니다~😊😊

yoonsseo commented on README.md:
```diff
@@ -1,2 +1,270 @@
+  3. 모든 Post 객체 조회
+<img width="730" alt="스크린샷 2023-09-30 오후 10 25 47" src="https://github.com/jongmee/spring-daagn-market-18th/assets/101439796/c8c0870a-5722-49cf-bbdb-2cd550d2288b">
+
+#### ✦ 추가 스터디 과제 질문들<br>
```
와 추가질문도 찾아보셨군용 덕분에 많이 알아가욥!!👏👏

yoonsseo reviewed (commented):
연휴동안 과제 수고하셨습니당🙂🙂

jongmee committed:
:memo: DB 설계 시 고려사항과 빌더 패턴 장점 리드미에 추가하기
